### PR TITLE
Configure ChoicePropertyEditor choices through annotations

### PR DESCRIPTION
**Pull request by jerry73204**
Currently AI2 is lack of a configurable choice-like property. It could be troublesome when I was developing components or extensions. It's a hassle to create editor classes for choice property classes (eg. `YoungAndroidLegoEv3{Color,Gyro,Ultrasonic}SensorModeChoicePropertyEditor`), and it seems there's no way to create choice-like properties for extensions without patching AI2 sources.

I made some modifications on ChoicePropertyEditor, and thus you can configure your choices with `editorArgs` argument in annotation like this.
```java
@DesignerProperty(editorType = PropertyTypeConstants.PROPERTY_TYPE_CHOICES,
                  defaultValue = "A",
                  editorArgs = {"A", "B", "C", "D"})
@SimpleProperty
public void SomeMethod(String choice) { /* blablabla */ }
```

This patch is mainly done by an additional String array `editorArgs` in class `ChoicePropertyEditor`. The `findComponentProperties` function passes `editorArgs` to editor objects if need.

### CODE REVIEW

**Comment by moliata** [comment]
This is a really great addition, thanks for your contribution. @ewpatton @jisqyv @josmas @afmckinney @paulmedwal please take a look at this.

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/ComponentDatabase.java [hunk 1]
```diff
@@ -331,9 +331,15 @@ private boolean initComponent(JSONObject componentNode) {
   private void findComponentProperties(ComponentDefinition component, JSONArray propertiesArray) {
     for (JSONValue propertyValue : propertiesArray.getElements()) {
       Map<String, JSONValue> properties = propertyValue.asObject().getProperties();
+
+      List<String> editorArgsList = new ArrayList<String>();
+      for (JSONValue val : properties.get("editorArgs").asArray().getElements())
```
Comment (jisqyv) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/ComponentDatabase.java]: This line causes a fault at runtime when loading a project. Basically editorArgs do not exist on existing components and are not present in existing projects. You should either deal with the case where editorArgs doesn't exist, or write a project upgrader that makes sure it does!

File: appinventor/components/src/com/google/appinventor/components/scripts/ComponentDescriptorGenerator.java [hunk 2]
```diff
@@ -137,7 +137,14 @@ private void outputProperty(String propertyName, DesignerProperty dp, StringBuil
...
     sb.append("\", \"defaultValue\": \"");
     sb.append(dp.defaultValue().replace("\"", "\\\""));
-    sb.append("\"}");
+
+    sb.append("\", \"editorArgs\": ");
+    String[] editorArgs = dp.editorArgs();
+    for (int idx = 0; idx < editorArgs.length; idx += 1)
+      editorArgs[idx] = "\"" + editorArgs[idx].replace("\"", "\\\"") + "\"";
+
+    sb.append("[" + String.join(", ", editorArgs) + "]");
```
Comment (jisqyv) [inline appinventor/components/src/com/google/appinventor/components/scripts/ComponentDescriptorGenerator.java]: Production compiles are still performed with Java 7. Java 7 doesn't have a String.join method.

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/ComponentDatabase.java (hunk 1 above)
Comment (jerry73204) [reply appinventor/appengine/src/com/google/appinventor/client/editor/simple/ComponentDatabase.java]: Sorry, I missed this point. Can you give me some guidance on writing a project upgrader?

**Comment by jerry73204** [comment]
String.join is now replaced with a StringBuilder implementation.

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/ComponentDatabase.java (hunk 1 above)
Comment (jerry73204) [reply appinventor/appengine/src/com/google/appinventor/client/editor/simple/ComponentDatabase.java]: Currently I cannot reproduce the bug. I creates some .aia files from MIT official server. My server ([link](http://wtf.csie.org:8888/)) imports these files without error.

Here are the configs for my build.
* ArchLinux with OpenJDK 7 as default.
* Compile with `ant clean && ant -Ddebug=true`.
* Start the server with `dev_appserver.sh --jvm_flag=-Xdebug --jvm_flag=-Xrunjdwp:transport=dt_socket,server=y,suspend=n,address=55555 --address=0.0.0.0 --port=8888 appengine/build/war` and `ant RunLocalBuildServer`.

**Comment by jerry73204** [comment]
An workaround that checks if editorArgs exists has been uploaded. I leave a comment above this code saying that it may need better solution in future versions. (something like component upgrader)

**Comment by ewpatton** [comment]
@jisqyv Do you think we can get this in as part of the Dec 20 release?

**Comment by jisqyv** [comment]
@jerry73204 LGTM. This code has been squashed and moved to our Gerrit system. It will be merged from there and I will close this PR when that happens (really soon, this code is scheduled to be released in the next component release which is scheduled for 12/27/2017)

**Comment by jisqyv** [comment]
@jerry73204 Code merged into the ucr branch

**Pull request closed without merging**